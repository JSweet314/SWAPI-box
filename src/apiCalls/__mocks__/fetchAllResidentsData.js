/* eslint-disable no-undef */
export const fetchAllResidentsData = jest.fn();
/* eslint-enable no-undef */