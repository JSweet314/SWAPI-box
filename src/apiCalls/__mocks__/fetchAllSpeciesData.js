/* eslint-disable no-undef */
export const fetchAllSpeciesData = jest.fn();
/* eslint-enable no-undef */