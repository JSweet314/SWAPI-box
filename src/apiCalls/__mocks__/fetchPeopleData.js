/* eslint-disable no-undef */
export const fetchPeopleData = jest.fn();
/* eslint-enable no-undef */