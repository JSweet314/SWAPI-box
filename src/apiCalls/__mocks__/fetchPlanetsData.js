/* eslint-disable no-undef */
export const fetchPlanetsData = jest.fn();
/* eslint-enable no-undef */