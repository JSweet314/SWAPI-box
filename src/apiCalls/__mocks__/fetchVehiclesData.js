/* eslint-disable no-undef */
export const fetchVehiclesData = jest.fn();
/* eslint-enable no-undef */