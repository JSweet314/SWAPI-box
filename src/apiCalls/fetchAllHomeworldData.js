export const fetchAllHomeworldData = (peopleData) => 
  Promise.all(peopleData.peopleArray.map(fetchHomeworld))
    .then(peopleArray => ({...peopleData, peopleArray}));

export const fetchHomeworld = person => 
  fetch(person.homeworld)
    .then(response => response.json())
    .then(wrangleHomeworldData)
    .then(planetData => ({ ...person, ...planetData }));

export const wrangleHomeworldData = planetData => 
  ({ homeworld: planetData.name, population: planetData.population });
