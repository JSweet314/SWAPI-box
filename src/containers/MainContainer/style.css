.main-display {
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  align-items: center;
}

.loading-display {
  width: 100%;
  height: 100%;
}

.loading-display img {
  width: 100%;
  height: 90%;
  opacity: .3;
}