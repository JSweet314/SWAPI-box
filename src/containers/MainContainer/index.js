import React, {Component} from 'react';
import Favorites from '../../components/Favorites/index';
import {fetchAllHomeworldData} from '../../apiCalls/fetchAllHomeworldData';
import {fetchAllResidentsData} from '../../apiCalls/fetchAllResidentsData';
import {fetchAllSpeciesData} from '../../apiCalls/fetchAllSpeciesData';
import {fetchPeopleData} from '../../apiCalls/fetchPeopleData';
import {fetchPlanetsData} from '../../apiCalls/fetchPlanetsData';
import {fetchVehiclesData} from '../../apiCalls/fetchVehiclesData';
import loadingGIF from '../../images/Loading_icon.gif';
import PageButtons from '../../components/PageButtons/index.js';
import PersonCard from '../../components/PersonCard/index';
import PlanetCard from '../../components/PlanetCard/index';
import PropTypes from 'prop-types';
import VehicleCard from '../../components/VehicleCard/index';
import './style.css';

export default class MainContainer extends Component {
  constructor(props) {
    super(props);
    this.state = {
      peopleArray: [],
      planetsArray: [],
      vehiclesArray: [],
      next: null,
      previous: null,
      pageNumber: 1,
      loading: true
    };
  }

  componentDidMount = () => {
    this.findClosestData();
  }

  componentDidUpdate = (prevProps) => {
    if (prevProps.match.params.id !== this.props.match.params.id) {
      this.findClosestData();
    }
  }

  findClosestData = () => {
    const priorData = localStorage.getItem(
      `SWAPI-${this.props.match.params.id}`
    );
    if (priorData) {
      const dataArray = `${this.props.match.params.id}Array`;
      const value = JSON.parse(priorData);
      this.setState({[dataArray]: value, loading: false});
    } else {
      this.setState({pageNumber: 1, loading: true}, this.getCategoryById);
    }
  }

  getCategoryById = () => {
    switch (this.props.match.params.id) {
    case 'people':
      this.getPeopleData();
      return;
    case 'planets':
      this.getPlanetsData();
      return;
    case 'vehicles':
      this.getVehiclesData();
      return;
    default:
      break;
    }
  }

  buildCards = () => {
    switch (this.props.match.params.id) {
    case 'people':
      return this.personCards();
    case 'planets':
      return this.planetCards();
    case 'vehicles':
      return this.vehicleCards();
    default:
      return null;
    }
  }

  getPeopleData = () => 
    fetchPeopleData()
      .then(fetchAllHomeworldData)
      .then(fetchAllSpeciesData)
      .then(this.deployPeopleData)
      .catch(error => alert(error.message));

  deployPeopleData = peopleData =>
    this.setState({
      next: peopleData.next,
      previous: peopleData.previous,
      peopleArray: peopleData.peopleArray,
      loading: false
    }, this.storeCategoryData)

  personCards = () => {
    const {handleOnClick, favorites} = this.props;
    return this.state.peopleArray.map(card =>
      <PersonCard
        handleOnClick={handleOnClick}
        favorites={favorites}
        key={card.name}
        card={card} />
    );
  }

  getPlanetsData = () => 
    fetchPlanetsData()
      .then(fetchAllResidentsData)
      .then(this.deployPlanetsData)
      .catch(error => alert(error.message));

  deployPlanetsData = planetsData => 
    this.setState({
      next: planetsData.next,
      previous: planetsData.previous,
      planetsArray: planetsData.planetsArray,
      loading: false
    }, this.storeCategoryData)

  planetCards = () => {
    const {handleOnClick, favorites} = this.props;
    return this.state.planetsArray.map(card =>
      <PlanetCard
        handleOnClick={handleOnClick}
        favorites={favorites}
        category="Planets"
        key={card.name}
        card={card} />
    );
  }

  getVehiclesData = () => 
    fetchVehiclesData()
      .then(this.deployVehiclesData)
      .catch(error => alert(error.message));

  deployVehiclesData = vehiclesData => 
    this.setState({
      next: vehiclesData.next,
      previous: vehiclesData.previous,
      vehiclesArray: vehiclesData.vehiclesArray,
      loading: false
    }, this.storeCategoryData)

  vehicleCards = () => {
    const {handleOnClick, favorites} = this.props;
    return this.state.vehiclesArray.map(card =>
      <VehicleCard
        handleOnClick={handleOnClick}
        favorites={favorites}
        key={card.name}
        card={card} />
    );
  }

  storeCategoryData = () =>
    localStorage.setItem(
      `SWAPI-${this.props.match.params.id}`, 
      JSON.stringify(this.state[`${this.props.match.params.id}Array`])
    );
   
  render = () => {
    const {handleOnClick, favorites} = this.props;
    const cards = this.buildCards();
    const {pageNumber} = this.state;
    if (this.props.match.params.id === 'favorites') {
      return <Favorites handleOnClick={handleOnClick} favorites={favorites} />;
    }

    return !this.state.loading ? 
      (
        <div className="people-display">
          {cards}
          <PageButtons pageNumber={pageNumber}/>
        </div>
      ) : (
        <div className="category-display">
          <img src={loadingGIF} alt="loading" />
        </div>
      );
  }
}

MainContainer.propTypes = {
  favorites: PropTypes.array.isRequired,
  handleOnClick: PropTypes.func.isRequired,
  match: PropTypes.shape({
    isExact: PropTypes.bool.isRequired,
    params: PropTypes.object.isRequired,
    path: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired
  })
};