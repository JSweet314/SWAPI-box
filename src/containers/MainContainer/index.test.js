import React from 'react';
import {shallow} from 'enzyme';
import MainContainer from './index';
import LocalStorage from '../../__mocks__/localStorageMock';
import {fetchAllHomeworldData} from '../../apiCalls/fetchAllHomeworldData';
import {fetchAllResidentsData} from '../../apiCalls/fetchAllResidentsData';
import {fetchAllSpeciesData} from '../../apiCalls/fetchAllSpeciesData';
import {fetchPeopleData} from '../../apiCalls/fetchPeopleData';
import {fetchPlanetsData} from '../../apiCalls/fetchPlanetsData';
import {fetchVehiclesData} from '../../apiCalls/fetchVehiclesData';

window.localStorage = new LocalStorage();

describe('MainContainer', () => {
  /*eslint-disable no-undef*/
  let wrapper;
  const mockGetDataByRouteId = jest.fn();
  const mockGetPeopleData = jest.fn();
  const mockGetPlanetsData = jest.fn();
  const mockGetVehiclesData = jest.fn();
  const mockProps = {
    favorites: [],
    handleFavoriteClick: jest.fn(),
    match: {
      isExact: false,
      params: {id: 'people'},
      path: "/category/:id",
      url: "/category/people"
    }
  };
  const mockPropsPlanetsRoute = {
    favorites: [],
    handleFavoriteClick: jest.fn(),
    match: {
      isExact: false,
      params: { id: 'planets' },
      path: "/category/:id",
      url: "/category/people"
    }
  };
  const mockPropsVehiclesRoute = {
    favorites: [],
    handleFavoriteClick: jest.fn(),
    match: {
      isExact: false,
      params: { id: 'vehicles' },
      path: "/category/:id",
      url: "/category/people"
    }
  };
  const mockPersonCard = {
    homeworld: "Tatooine",
    name: "Luke Skywalker",
    population: "200000",
    species: "Human"
  };
  const mockStoragePeople = {
    next: "https://swapi.co/api/people/?format=json&page=2",
    previous: null,
    dataArray: [mockPersonCard]
  };
  /*eslint-enable no-undef*/
  beforeEach(() => {
    localStorage.clear();
    wrapper = shallow(
      <MainContainer {...mockProps}/>,
      {disableLifecycleMethods: true}
    );
  });

  it('should match a snapshot with no data present', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should match a snapshot with data present', () => {
    wrapper.setState({
      people: [mockPersonCard],
      loading: false
    });
    expect(wrapper).toMatchSnapshot();
  });

  describe('findClosestData', ()=>{
    test('should pull from localStorage if possible', () => {
      localStorage.setItem('SWAPI-people', JSON.stringify(mockStoragePeople));
      expect(wrapper.state('people')).toEqual([]);
      wrapper.instance().findClosestData();
      expect(wrapper.state('people')).toEqual([mockPersonCard]);
    });

    it('should call a function if localStorage is not possible', 
      () => {
        wrapper.instance().getDataByRouteId = mockGetDataByRouteId;
        wrapper.instance().findClosestData();
        expect(mockGetDataByRouteId).toHaveBeenCalled();
      }
    );

    it('sets loading to true in state if a fetch in needed',
      () => {
        expect(wrapper.state('loading')).toEqual(false);
        wrapper.instance().getDataByRouteId = mockGetDataByRouteId;
        wrapper.instance().findClosestData();
        expect(wrapper.state('loading')).toEqual(true);
      }
    ); 
  });

  describe('getDataByRouteId', () => {
    it('should call another function based on route id', () => {
      wrapper.instance().getPeopleData = mockGetPeopleData;
      wrapper.instance().getDataByRouteId();
      expect(mockGetPeopleData).toHaveBeenCalled();
    });

    it('should be able to call a function for planets route', () => {
      wrapper = shallow(
        <MainContainer {...mockPropsPlanetsRoute} />,
        { disableLifecycleMethods: true }
      );
      wrapper.instance().getPlanetsData = mockGetPlanetsData;
      wrapper.instance().getDataByRouteId();
      expect(mockGetPlanetsData).toHaveBeenCalled();
    });

    it('should be able to call a function for vehicles route', () => {
      wrapper = shallow(
        <MainContainer {...mockPropsVehiclesRoute} />,
        { disableLifecycleMethods: true }
      );
      wrapper.instance().getVehiclesData = mockGetVehiclesData;
      wrapper.instance().getDataByRouteId();
      expect(mockGetVehiclesData).toHaveBeenCalled();
    });
  });

  describe('getPeopleData', () => {});
  describe('deployPeopleData', () => {});
  describe('getPlanetsData', () => {});
  describe('deployPlanetsData', () => {});
  describe('getVehiclesData', () => {});
  describe('deployVehiclesData', () => {});


});