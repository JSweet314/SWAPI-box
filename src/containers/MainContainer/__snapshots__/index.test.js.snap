// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainContainer should match a snapshot with data present 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MainContainer
    favorites={Array []}
    handleFavoriteClick={[Function]}
    match={
        Object {
            "isExact": false,
            "params": Object {
              "id": "people",
            },
            "path": "/category/:id",
            "url": "/category/people",
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        Array [
          <PersonCard
            card={
                        Object {
                                    "homeworld": "Tatooine",
                                    "name": "Luke Skywalker",
                                    "population": "200000",
                                    "species": "Human",
                                  }
            }
            favorites={Array []}
            handleFavoriteClick={[Function]}
/>,
        ],
        <PageButtons
          handlePageButtonClick={[Function]}
          next={null}
          previous={null}
/>,
      ],
      "className": "people-display",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "Luke Skywalker",
        "nodeType": "function",
        "props": Object {
          "card": Object {
            "homeworld": "Tatooine",
            "name": "Luke Skywalker",
            "population": "200000",
            "species": "Human",
          },
          "favorites": Array [],
          "handleFavoriteClick": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "handlePageButtonClick": [Function],
          "next": null,
          "previous": null,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          Array [
            <PersonCard
              card={
                            Object {
                                          "homeworld": "Tatooine",
                                          "name": "Luke Skywalker",
                                          "population": "200000",
                                          "species": "Human",
                                        }
              }
              favorites={Array []}
              handleFavoriteClick={[Function]}
/>,
          ],
          <PageButtons
            handlePageButtonClick={[Function]}
            next={null}
            previous={null}
/>,
        ],
        "className": "people-display",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "Luke Skywalker",
          "nodeType": "function",
          "props": Object {
            "card": Object {
              "homeworld": "Tatooine",
              "name": "Luke Skywalker",
              "population": "200000",
              "species": "Human",
            },
            "favorites": Array [],
            "handleFavoriteClick": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "handlePageButtonClick": [Function],
            "next": null,
            "previous": null,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;

exports[`MainContainer should match a snapshot with no data present 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MainContainer
    favorites={Array []}
    handleFavoriteClick={[Function]}
    match={
        Object {
            "isExact": false,
            "params": Object {
              "id": "people",
            },
            "path": "/category/:id",
            "url": "/category/people",
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        Array [],
        <PageButtons
          handlePageButtonClick={[Function]}
          next={null}
          previous={null}
/>,
      ],
      "className": "people-display",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "handlePageButtonClick": [Function],
          "next": null,
          "previous": null,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          Array [],
          <PageButtons
            handlePageButtonClick={[Function]}
            next={null}
            previous={null}
/>,
        ],
        "className": "people-display",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "handlePageButtonClick": [Function],
            "next": null,
            "previous": null,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
